google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.api.core.ApiAsyncFunction;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.BetaApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.ApiStreamObserver;
+import com.google.cloud.firestore.Query.QueryOptions;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.grpc.Status;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Class used to store state required for running a recursive delete operation. Each recursive
+ * delete call should use a new instance of the class.
+ */
+@BetaApi
+public final class RecursiveDelete {
+  /**
+   * Datastore allowed numeric IDs where Firestore only allows strings. Numeric IDs are exposed to
+   * Firestore as __idNUM__, so this is the lowest possible negative numeric value expressed in that
+   * format.
+   *
+   * <p>This constant is used to specify startAt/endAt values when querying for all descendants in a
+   * single collection.
+   */
+  public static final String REFERENCE_NAME_MIN_ID = "__id-9223372036854775808__";
+
+  /**
+   * The query limit used for recursive deletes when fetching all descendants of the specified
+   * reference to delete. This is done to prevent the query stream from streaming documents faster
+   * than Firestore can delete.
+   */
+  public static final int MAX_PENDING_OPS = 5000;
+
+  /**
+   * The number of pending BulkWriter operations at which RecursiveDelete starts the next limit
+   * query to fetch descendants. By starting the query while there are pending operations, Firestore
+   * can improve BulkWriter throughput. This helps prevent BulkWriter from idling while Firestore
+   * fetches the next query.
+   */
+  public static final int MIN_PENDING_OPS = 1000;
+
+  private final FirestoreRpcContext<?> firestoreRpcContext;
+  private final BulkWriter writer;
+
+  @Nullable private final DocumentReference documentReference;
```

Comment:
It might be cleaner to replace this with a documentId string. Then you would have parentPath, collectionId and documentId below. What do you think?

Reply:
I think trying to rehydrate the DocumentReference from the documentId is more bug-prone, since we'd be setting it to an empty string for collections, and the document path for documents.

I renamed `documentReference` to `rootDocumentReference` and added documentation. This way, we don't have to add extra code to rehydrate, and it's pretty clear to see the purpose of the variable at a glance.